### ISSUE TITLE
False negative of potentially Optional type inside lambda body

### ISSUE THREAD

**Issue by pawelrubin** [microsoft/pyright#1669]
**To Reproduce**
Consider the following piece of code:

```python
def eratosthenes_stream(n: int) -> Iterator[int]:
    def sieve(stream: Iterator[int]) -> Iterator[int]:
        if head := next(stream, None):
            yield head
            yield from sieve(x for x in stream if x % head != 0)

    return sieve(iter(range(2, n + 1)))
```

This works just fine and raises no type errors both in **PyRight** and **mypy**. However, if the 
`sieve` will be called with `filter` with a `lambda` inside, it generates errors in both type checkers.

```python
def eratosthenes_stream(n: int) -> Iterator[int]:
    def sieve(stream: Iterator[int]) -> Iterator[int]:
        if head := next(stream, None):
            yield head
            yield from sieve(filter(lambda x: x % head != 0, stream))
			#                                 ^^^^^^^^^^^^^
            # Operator "%" not supported for types "int" and "int | None"
            # Return type of lambda is unknown

    return sieve(iter(range(2, n + 1)))
```

I have narrowed the code to a simple case with just a lambda:

```python
def foo(x: Optional[int]) -> Iterator[int]:
    if x is not None:
        yield (lambda: x + 1)()
        #              ^
        # Operator "+" not supported for "None"
```
And it still reports the error.

It seems to be an issue connected to lambdas only.

**Expected behavior**
The filter-lambda approach should not report any errors and the type of `head` inside the `lambda` function should be revealed as `int` not `int | None`.

**Additional context**

- Python 3.9.1 64-bit
- Mac OS Big Sur 11.2.3
- Pylance v2021.3.2

***VS Code***
Version: 1.54.3
Commit: 2b9aebd5354a3629c3aba0a5f5df49f43d6689f8
Date: 2021-03-15T11:57:12.728Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 20.3.0

**Comment by erictraut** [comment]
This behavior is intended. A type checker does not know when a lambda is going to be executed, so it cannot assume that a captured variable's type will not be modified between the time the lambda is passed to another function and it is called. For more details, refer to this issue: https://github.com/microsoft/pyright/issues/599.

**Comment by pawelrubin** [comment]
I get it now. It would be very hard to prove the narrowed type of a variable in the general case. 

I'll avoid using lambdas whenever possible, and I'll stick to the generator-comprehension approach in this case.

By the way, thank you for your awesome work on Pyright! This is an incredible project and a key tool in my daily workflow! 🎖️

**Comment by erictraut** [comment]
I found a way to reliably apply type narrowing to a local variable or parameter that is captured by an inner-scoped lambda or function. The narrowed type is allowed only in cases where the type analyzer can prove that there is no code path that assigns a new value to the symbol after the lambda or def statement.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.172, which I just published. It will also be in the next release of pylance.
